A58-client-side-weighted-round-robin-lb-policy.md

```diff
@@ -28,42 +29,56 @@ Introduce a new LB policy `weighted_round_robin`. This policy implements client-
+  // value, if we later start seeing fresh reports again, the
+  // blackout_period applies.  Defaults to 3 minutes.
+  google.protobuf.Duration weight_expiration_period = 4;
+
+  // How often endpoint weights are recalculated.  Default is 1 second.
+  google.protobuf.Duration weight_update_period = 5;
 }
-
 ```
 
-`blackout_period` is to handle backend churns. `weighted_round_robin` rejects load reports until a backend reports valid weights for at least `blackout_period` during which the average weight value is used instead. Disables blackout when set to a negative value. Defaults to 10 seconds.
-
-Load metrics for a subchannel expire after `weight_expiration_period` is passed without any update from the backend, including with per-call or OOB load reporting. Once expired, the average weight value is used instead. Defaults to 3 minutes. Cannot be disabled but the period has no max value allowed.
-
-The client initiates OOB load reporting by sending a request only when `enable_oob_load_report` is set. Defaults to false. When set, this policy does not parse per-call load reports. See below for more on load reporting. Note this does not affect the per-call load reports in xDS LRS.
-
-`oob_reporting_period` is only used when OOB load reporting is used. Note the effective reporting period is determined as outlined in [gRFC A51] and may be different. The client requests the minimum of values from all interested LB policies, but the server enforces its own minimum reporting period. Defaults to 10 seconds.
-
-The client updates effective weights periodically every `weight_update_period`. This parameter defaults to 1 second. The lowest value allowed is 100ms. When set to a value lower than 100ms the period is set to 100ms. Weight updates triggered by changes in the connectivity state or backend addresses are not affected.
-
 ### Earliest Deadline First (EDF) Scheduler
 
-`weighted_round_robin` introduces a scheduler that manages weighted subchannel selection. A new scheduler instance is created every time weights are updated.  A state object is passed when created so that the scheduler can start where it last left across multiple instances. The scheduler must be light and cheap to create and allow concurrent pick calls. The scheduler returns the index in the input weight array and the caller has to map the index to a subchannel.
+`weighted_round_robin` introduces a scheduler that manages weighted subchannel selection. A new scheduler will be created periodically to take new weight information into account. A state object is passed when created so that the scheduler can start where it last left across multiple instances. The scheduler must be light and cheap to create and allow concurrent pick calls. The scheduler returns the index in the input weight array and the caller has to map the index to a subchannel. The scheduler's state is stored in the LB policy and not the picker, and is atomically updated at the end of each pick by the scheduler.
```

Comment:
This doesn't say anything about how this state object is actually used.  What value is stored in the state object?  How is it used and updated by the scheduler?

Also, have we resolved whether we want all languages to use this approach, or are we okay with having different languages use different approaches?  I thought we said that our internal implementation does not use this atomic state approach in Java or Go.

Reply:
Updated after chatting with Ivy and Erik.

I removed anything specific to the static stride and the global state object (next_seq lambda). Java will for now use a priority queue, randomize the starting index with each scheduler, and hold a lock in pick() (Java stubby implementation).